### PR TITLE
Correctly capturing and reporting Node errors.

### PR DESCRIPTION
**Pull request by jnwng**
Errors that occur outside of the test harness (most specifically, due to Node errors like misconfigured `require` statements) would sometimes falsely marked tests as "failed" based on the distribution of test files to various worker processes.

Additionally, because Jenkins does not mark builds as unstable or failed based on aforementioned Node errors, I'm capturing all Node errors and reporting them as test failures corresponding to the file they originated from. 

There's a lot of repeated code in the files I touched, I refactored them as best I could to make sure we weren't repeating logic everywhere.

---

Why is it important to report these errors instead of just letting them bubble up and kill the process (and subsequently causing a build failure)? I think there's some value in having the test runners run through all the tests and make sure we can capture all possible failures in one run, instead of having one test obscure the results of the others.

### CODE REVIEW

**Comment by jnwng** [comment]
@eleith PTAL.

unfortunately, it looks like something i've changed here resulted in the test runner only capturing a fraction of tests. any ideas?

File: lib/mocha.js [hunk 1]
```diff
@@ -23,6 +23,53 @@ var overrideBddIt = function(context, file, mocha) {
...
+        process.send({status:'pass', file:test.file, title:test.title});
+      }
+    });
+  } catch (err) {
+    var filename = mocha.files.length > 1 ? 'Multiple tests failed' : mocha.files[0];
+    // When we have node.js errors, we need to report them differently.
+    if (process && process.send) {
+      console.log(err.stack);
+      process.send({status: 'error', file: filename, message: err.message, stack: err.stack});
+      process.send({status: 'end'});
```
Comment (eleith) [inline lib/mocha.js]: why are you sending an 'end' here?

Comment (eleith) [inline lib/mocha.js]: oh, i assume that the callback to 'run' never gets called on an error?

File: lib/mocha.js [hunk 2]
```diff
@@ -49,75 +96,23 @@ exports.runner = function(config, configFile) {
...
-        process.send({status:'fail', file:test.file, title:test.title, message:err.message});
-      }
-    });
-
-    runner.on('pass', function(test){
-      if (process && process.send) {
-        process.send({status:'pass', file:test.file, title:test.title});
-      }
-    });
+    runMocha(mocha, testFiles[0]);
```
Comment (eleith) [inline lib/mocha.js]: why pass in testFiles[0] ?

File: lib/mocha.js [hunk 3]
```diff
@@ -23,6 +23,53 @@ var overrideBddIt = function(context, file, mocha) {
...
+  return mocha;
+}
+
+function runMocha(mocha, testFile) {
+  try {
+    var runner = mocha.run(function(failures) {
+      if (process && process.send) {
+        process.send({status:'end', file: testFile});
+      }
+      process.exit(failures);
```
Comment (eleith) [inline lib/mocha.js]: this is different for unify vs non unify

**Comment by eleith** [comment]
lgtm

**Pull request merged**